controllers/openstackdataplanenodeset_controller.go

```diff
@@ -316,28 +328,117 @@ func (r *OpenStackDataPlaneNodeSetReconciler) Reconcile(ctx context.Context, req
 	return ctrl.Result{}, nil
 }
 
-func checkDeployment(helper *helper.Helper,
-	request ctrl.Request) (bool, bool, error) {
+func checkDeployment(
+	ctx context.Context,
+	helper *helper.Helper,
+	instance *dataplanev1.OpenStackDataPlaneNodeSet,
+	configMapHashes map[string]string,
+	secretHashes map[string]string,
+) (bool, bool, error) {
+
 	// Get all completed deployments
 	deployments := &dataplanev1.OpenStackDataPlaneDeploymentList{}
 	opts := []client.ListOption{
-		client.InNamespace(request.NamespacedName.Namespace),
+		client.InNamespace(instance.Namespace),
 	}
-	err := helper.GetClient().List(context.Background(), deployments, opts...)
+	err := helper.GetClient().List(ctx, deployments, opts...)
 	if err != nil {
 		helper.GetLogger().Error(err, "Unable to retrieve OpenStackDataPlaneDeployment CRs %v")
 		return false, false, err
 	}
+
+	hashNodeSetServices := false
+	deploymentExists := false
+	deploymentReady := false
+
 	for _, deployment := range deployments.Items {
 		if slices.Contains(
-			deployment.Spec.NodeSets, request.NamespacedName.Name) {
+			deployment.Spec.NodeSets, instance.Name) {
+
+			deploymentExists = true
+
 			if deployment.Status.Deployed {
-				return true, true, nil
+				deploymentReady = true
+				if len(deployment.Spec.ServicesOverride) > 0 {
+					for _, serviceName := range deployment.Spec.ServicesOverride {
+						err = getServiceHashes(ctx, helper, instance, serviceName, configMapHashes, secretHashes)
```

Comment:
Does it make sure that we record the hash of the CM/Secrets at the time when the Deployment was done? E.g. if the CM/Secret is updated after the Deployment became Ready then we should not put the new hash of the CM/Secret to the NodeSet.

Reply:
That is a good point. The time delta when such a change could occur would be from the CM/Secret is mounted into the AnsibleEE pod, and when the Deployment.Status.Deployed is set to true, which triggers the watch reconcile for the NodeSet. Given there could be ansible executions happening in between there, the delta could be large.

To mitigate that, I can update this PR to also save the hashes on the Deployment Status as well, as they are used. The NodeSet would then just read and copy the hash values from the Deployment. There is still a race condition there in that the CM/Secrets could change, but it's a much smaller window.

Reply:
I think we need to persist the hash of the CM/Secret when the Deployment first becomes Ready. If we do so then we have a snapshot of the hash from the time of the Deployment. Then the nodeset can copy the hash from the newest Deployment to show what is deployed on the nodes.  I think we are on the same page about this. 

As you noted there is still a race window but smaller. If we want to make this race free then we need to prevent CM/Secret changing while the deployment is running. One way to do that would be to use immutable CM/Secret objects https://kubernetes.io/docs/concepts/configuration/configmap/#configmap-immutable However that would need multiple changes:
* service operators would need to generate immutable CM/Secrets instead of updating them in place
* the user would need to update the DataplaneService definition to point to the new CM/Secret object after the service operator generated the new inputs.

How do you feel, do we want to go towards immutable CM/Secret usage?

Reply:
In the latest iteration, I've moved the calculation of the hashes to be done when the Deployment is completed and stored on the Deployment.Status. When the NodeSet Deployment watcher triggers a reconcile, they'll be copied to NodeSet.Status. I still need to add more tests and update the docs to account for this change. I'll mark this WIP until those are done.  However, in the meantime let me know if you think this approach is better.

I think we should defer the discussion around immutable ConfigMaps/Secrets to a separate effort, unless you think that doing that would make the hash calculation obsolete.

Reply:
> I think we should defer the discussion around immutable ConfigMaps/Secrets to a separate effort, unless you think that doing that would make the hash calculation obsolete.

Unfortunately that does not make the hash calculation obsolete as the DataPlaneService CRs are mutable and therefore can be changed to point to new CMs/Secrets. So the immutable CM/Secret only helps with the race window. So I guess we would not gain that much. Let's park this part of the discussion.

> In the latest iteration, I've moved the calculation of the hashes to be done when the Deployment is completed and stored on the Deployment.Status. When the NodeSet Deployment watcher triggers a reconcile, they'll be copied to NodeSet.Status. I still need to add more tests and update the docs to account for this change. I'll mark this WIP until those are done. However, in the meantime let me know if you think this approach is better.

This sounds good to me but I have to look at the code too.